gnrc_netif: centralize device-type-specific functions

miri64 opened the pull request:
### Contribution description
This moves all remaining  `switch (netif->device_type)`-like functions to `gnrc_netif_device_type.c` and also applies the warning mechanisms introduced in #10513. The hope is that if a new device type is implemented the implementor will stumble over these functions basically immediately.

### Testing procedure
Pinging a global address with `gnrc_networking` via a `gnrc_border_router` should still work for

- [x] Ethernet (here just two `gnrc_networking` examples with one having `ifconfig <if> -rtr_adv` and the other a global address configured by hand should be enough)
- [x] IEEE 802.15.4
- [ ] Nordic BLE softdevice
- [ ] ~~cc110x~~ (see https://github.com/RIOT-OS/RIOT/issues/10723)
- [ ] ~~nrfmin~~ (see https://github.com/RIOT-OS/RIOT/issues/10723)
- [ ] ~~ESP NOW~~ (see https://github.com/RIOT-OS/RIOT/issues/10723)

After pinging the global address of the other node, the neighbor cache (`nib neigh`) of the border router (or advertising router - so the one were you *didn't* type `ifconfig <if> -rtr_adv`) should contain the other node's global address with the correct link-layer address

### Issues/PRs references
Depends on ~~#10513~~ (merged)

![gnrc_netif/gnrc_sixlowpan_iphc BLE capability](http://page.mi.fu-berlin.de/mlenders/netif_ble.svg)

miri64 commented:
Rebased and adapted for current #10513.

miri64 commented on sys/net/gnrc/netif/gnrc_netif_device_type.c:
```diff
@@ -0,0 +1,298 @@
+#if defined(MODULE_CC110X)
+        case NETDEV_TYPE_CC110X:
+            (void)opt;
+            return sizeof(uint8_t);
```
I added this (in an extra commit), as it wasn't implemented yet.

miri64 commented:
Rebased to current master. No longer dependent on another PR.

miri64 commented:
Rebased again and squashed this time

gschorcht commented:
@miri64 Tested. 

After configuring `2001:db8::1` as global unicast address on the node with enabled `RTR_ADV`, the other node with disabled `RTR_ADV` gets the global unicast address `2001:db8::32ae:a4ff:fe18:7a3d`. `nib neigh` on this gives:
```
> nib neigh
fe80::32ae:a4ff:fe41:60f9 dev #9 lladdr 30:AE:A4:41:60:F9 router REACHABLE
2001:db8::1 dev #9 lladdr 30:AE:A4:41:60:F9  REACHABLE
```

After pinging any global unicast address, `nib neigh` on the node with enabled `RTR_ADV` gives:
```
> nib neigh
fe80::32ae:a4ff:fe18:7a3d dev #9 lladdr 30:AE:A4:18:7A:3D router DELAY
2001:db8::32ae:a4ff:fe18:7a3d dev #9 lladdr 30:AE:A4:18:7A:3D  REACHABLE
```
I guess that is what you expected.

miri64 commented:
Yepp :-)

miri64 commented:
Rebased and adapted for current master.

miri64 commented:
ESP-now needs retesting, since I needed to adopt for #10581.

gschorcht commented:
> ESP-now needs retesting, since I needed to adopt for #10581.

Hm, not working in the first test. I will have to enable debugging, to see the router advertisements. Do I have to change the test procedure?

miri64 commented:
Ah yes, since it is now a 6Lo device you might need a `gnrc_border_router` on the second node.

gschorcht commented:
> Ah yes, since it is now a 6Lo device you might need a `gnrc_border_router` on the second node.

This is what I thought. So I tried to build a border router, but `ethos` seems not to work properly with esp32. Although the communication seems to work, the `help` command doesn't show any command that could be used.

Therefore, I tried also to build the boarder router with `esp_eth`. Unfortunately, `arb` is `NULL` here https://github.com/RIOT-OS/RIOT/blob/a971f29149ebe6c576e420439db3b9b34dfebe4f/sys/net/gnrc/network_layer/ipv6/nib/nib.c#L1444-L1447
when valid prefix information is received on the ethernet interface.
```
nib: received valid Prefix Information option:
     - Prefix: 2003:c1:e72f:6c40:52a7:2bff:fe7b:9caa/64
     - Flags: LA
     - Valid lifetime: 604800
     - Preferred lifetime: 86400
```
Have you any hint where to look for the problem?

gschorcht commented:
I have provided PR https://github.com/miri64/RIOT/pull/32 to fix this https://github.com/RIOT-OS/RIOT/blob/a971f29149ebe6c576e420439db3b9b34dfebe4f/sys/net/gnrc/netif/gnrc_netif_device_type.c#L139

Whith this fix, the `esp-now` nodes can ping each other again.

But the advertisement of global routing prefix on `esp_now` interface does not work, maybe because of disabled `GNRC_IPV6_NIB_CONF_MULTIHOP_P6C`.

miri64 commented:
I guess... Try it.

miri64 commented:
@maribu would be great if you could test cc110x as well.

gschorcht commented:
@miri64 When building a router with `esp_now` and `esp_eth`, I can't ping `esp_now` nodes anymore since 6Lo is not used for `esp_now` interface even though all required modules are enabled. It is simply trying to send IPv6 packets. How is it decided whether to use 6Lo for `esp_now` interface or not? I couldn't figure out the dependency. I just replaced `ethos` by `esp_now` in the makefile.

It is not a router problem, I also have this problem when I simply enable both interfaces.

miri64 commented:
> @miri64 When building a router with `esp_now` and `esp_eth`, I can't ping `esp_now` nodes anymore since 6Lo is not used for `esp_now` interface even though all required modules are enabled. It is simply trying to send IPv6 packets. How is it decided whether to use 6Lo for `esp_now` interface or not? I couldn't figure out the dependency. I just replaced `ethos` by `esp_now` in the makefile.

Shouldn't `ethos` be replaced with `esp_eth`? The defining function for 6Lo is here

https://github.com/RIOT-OS/RIOT/blob/9554f751d8a08979345544e220a5cd7feb709f8c/sys/include/net/gnrc/netif/internal.h#L342-L365
https://github.com/RIOT-OS/RIOT/blob/9554f751d8a08979345544e220a5cd7feb709f8c/sys/net/gnrc/netif/gnrc_netif.c#L1113-L1127

gschorcht commented:
> Shouldn't ethos be replaced with esp_eth? The defining function for 6Lo is here

Yes of course, I meant `esp_eth`, but I have to add both interfaces 
```
USEMODULE += esp_eth
USEMODULE += esp_now
```
.

gschorcht commented:
Hm, that's the problem, `NETDEV_TYPE_ESP_NOW` https://github.com/RIOT-OS/RIOT/blob/9554f751d8a08979345544e220a5cd7feb709f8c/sys/net/gnrc/netif/gnrc_netif.c#L1113-L1123
was merged out by 21261a8c306e20532fdecadfd3df4e9c8032a356

miri64 commented on sys/net/gnrc/netif/gnrc_netif_device_type.c:
```diff
@@ -17,11 +17,105 @@
+
+                res = dev->driver->get(dev, NETOPT_SRC_LEN, &tmp, sizeof(tmp));
+                assert(res == sizeof(tmp));
+                netif->l2addr_len = (uint8_t)tmp;
```
It did not cause any problems in current master (this code just was moved from `gnrc_netif.c`) and things have to be very wrong to get to this. But sure, why not?

miri64 replied:
Addressed

miri64 commented:
> Hm, that's the problem, `NETDEV_TYPE_ESP_NOW`
> 
> [RIOT/sys/net/gnrc/netif/gnrc_netif.c](https://github.com/RIOT-OS/RIOT/blob/9554f751d8a08979345544e220a5cd7feb709f8c/sys/net/gnrc/netif/gnrc_netif.c#L1113-L1123)
> 
> Lines 1113 to 1123 in [9554f75](/RIOT-OS/RIOT/commit/9554f751d8a08979345544e220a5cd7feb709f8c)
> 
>  #if (GNRC_NETIF_NUMOF > 1) || !defined(MODULE_GNRC_SIXLOWPAN) 
>  bool gnrc_netif_is_6ln(const gnrc_netif_t *netif) 
>  { 
>      switch (netif->device_type) { 
>          case NETDEV_TYPE_IEEE802154: 
>          case NETDEV_TYPE_CC110X: 
>          case NETDEV_TYPE_BLE: 
>          case NETDEV_TYPE_NRFMIN: 
>          case NETDEV_TYPE_ESP_NOW: 
>              return true; 
>          default: 
> 
> was merged out by [21261a8](https://github.com/RIOT-OS/RIOT/commit/21261a8c306e20532fdecadfd3df4e9c8032a356)

Oops, will add fixup!

miri64 commented:
@maribu @gschorcht addressed your comments.

gschorcht commented:
@miri64 Thanks :smile:

gschorcht commented:
@miri64 Since I have not yet seen how the test should look like if it works correctly on a platform, here what I did, what I could observe, and some questions I have:

- I have two nodes and one border router.
- The first question came up when I had to compile the nodes as leaf nodes with `CFLAGS="-DGNRC_IPV6_NIB_CONF_6LN=1 - DGNRC_IPV6_NIB_CONF_6LR=0`. Otherwise, I get
   
   ```
    nib: multihop prefix and context dissemination on router activated,
         but no ABRO found. Discarding router advertisement silently
   ```
   on receiption of router advertisements and the global unicast address is not set.
   Is that correct?
- On the nodes, `ifconfig` shows that `RTR_ADV` option is set even though they are leaf nodes. Is that correct and do I have to execute `ifconfig 9 -rtr_adv`?
- In my network, the border router gets over its Ethernet interface the global routing prefix `2003:aaaa:bbbb:cccc::/64` from the LAN and the global unicast address `2003:aaaa:bbbb:cccc:32ae:a4ff:fe47:4ccd`.
- When I use `ifconfig 9 add unicast 2001:db8::1/64`, the leaf nodes get the global unicast addresses `2001:db8::32ae:a4ff:fe18:7a3d` and `2001:db8::32ae:a4ff:fe41:60f9`
- On router, `nib route` gives
   ```
   2003:aaaa:bbbb:cccc::/64 dev #6
   2001:db8::/64 dev #9
   default* via fe80::1 dev #6
   ```
- On the nodes, `nib route` gives
   ```
   default* via fe80::32ae:a4ff:fe47:4ccd dev #9
   ```
- Pinging from one node the other node with `ping6 2001:db8::32ae:a4ff:fe18:7a3d` doesnt work. The reason seems to be that the source node is using rather its link local address than its global unicast address. On the route, I can see the following message:

   ```
   ipv6: Received (src = fe80::32ae:a4ff:fe41:60f9, dst = 2001:db8::32ae:a4ff:fe18:7a3d, next header = 58, length = 12)
   ```
   So it is probably the a problem of `esp_now`.

gschorcht commented:
@miri64 Please add ` || DOXYGEN` here https://github.com/RIOT-OS/RIOT/blob/3af28369a8d49d7e18a0766ceb93098dd42c712e/sys/include/net/gnrc/netif.h#L80 to get it documented.

gschorcht commented:
@miri64 After a few hours of work, I was able to get it up and running :rocket: But it's just a hack right now, and I would need your support to solve the problem.

I had the following problems:
- When I build a border router with `esp_eth` and `esp_now`, I had to use following defines explicitly to get it working:
    ```
    GNRC_IPV6_NIB_CONF_SLAAC=1
    GNRC_IPV6_NIB_CONF_MULTIHOP_P6C=0
    ```
- The nodes had to be build as leaf nodes `GNRC_IPV6_NIB_CONF_6LR=0`. Otherwise, I get
   
    ```
    nib: multihop prefix and context dissemination on router activated,
         but no ABRO found. Discarding router advertisement silently
   ```
   on receiption of router advertisements and the global unicast address is not set.
- `esp_now_netdev` is now a 6Lo node, but has no long address with 8 byte size. Therefore, ARO records could not be used in neighbor solicitations https://github.com/RIOT-OS/RIOT/blob/f7e524d18eb3794fc8f3b0357a4ce233204d4a56/sys/net/gnrc/network_layer/ipv6/nib/_nib-arsm.c#L60-L65.

I could solve the last problem by adding the option `NETOPT_ADDRESS_LONG` to `esp_now_netdev`. At the moment I'm just using the 6 byte MAC address extended by two 0-bytes. Checking received ARO records is still a problem https://github.com/RIOT-OS/RIOT/blob/f7e524d18eb3794fc8f3b0357a4ce233204d4a56/sys/net/gnrc/network_layer/ipv6/nib/_nib-6ln.c#L88 Function `_is_iface_eui64` insists on a size of 8 bytes https://github.com/RIOT-OS/RIOT/blob/f7e524d18eb3794fc8f3b0357a4ce233204d4a56/sys/net/gnrc/network_layer/ipv6/nib/_nib-6ln.c#L32-L37. I changed it for the moment as following
```diff
-    return (netif->l2addr_len == sizeof(eui64_t)) &&
+    return /* (netif->l2addr_len == sizeof(eui64_t)) && */
            (memcmp(&netif->l2addr, eui64, netif->l2addr_len) == 0);
```
but this is just a hack. Do you have any idea for a better solution?

With these changes the nodes can ping each other using the global unicast address and they can ping any machine in the LAN via the border router. Of course, the way back has to be set as route.

One little problem I could observe was that the bourder router doesn't provide router advertisements on a regular base. Therefore, the assignment of global unicast addresses did only work on startup of the nodes when they are sending router solicitation. Is that correct?

gschorcht commented:
@miri64 Maybe a clean solution would be to use `NETOPT_IPV6_IID` instead of `NETOPT_ADDRESS_LONG` for ARO generation https://github.com/RIOT-OS/RIOT/blob/f7e524d18eb3794fc8f3b0357a4ce233204d4a56/sys/net/gnrc/network_layer/ipv6/nib/_nib-arsm.c#L60 since the ARO should contain the EUI-64, and `NETOPT_IPV6_IID` always delivers the EUI-64 independent on the device.

Checking ARO should then also be done against the `NETOPT_IPV6_IID` in  https://github.com/RIOT-OS/RIOT/blob/f7e524d18eb3794fc8f3b0357a4ce233204d4a56/sys/net/gnrc/network_layer/ipv6/nib/_nib-6ln.c#L32-L37

miri64 commented:
Also, the IID is not the interfaces EUI-64 (it refers to the IEEE 802.15.4 long address), as the second U/L-Bit gets still flipped. Maybe we should deactivate 6Lo-ND for ESP-now, as it is clearly only specified for IEEE 802.15.4m see abstract in [RFC 6775](https://tools.ietf.org/html/rfc6775):

>    The IETF work in IPv6 over Low-power Wireless Personal Area Network
>    (6LoWPAN) defines 6LoWPANs such as IEEE 802.15.4.  This and other
>    similar link technologies have limited or no usage of multicast
>    signaling due to energy conservation.  In addition, the wireless
>    network may not strictly follow the traditional concept of IP subnets
>    and IP links.  IPv6 Neighbor Discovery was not designed for non-
>    transitive wireless links, as its reliance on the traditional IPv6
>    link concept and its heavy use of multicast make it inefficient and
>    sometimes impractical in a low-power and lossy network.  This
>    document describes simple optimizations to IPv6 Neighbor Discovery,
>    its addressing mechanisms, and duplicate address detection for Low-
>    power Wireless Personal Area Networks and similar networks.  The
>    document thus updates RFC 4944 to specify the use of the
>    optimizations defined here.

gschorcht commented:
> Sorry no time to read all your comments above right now (I will do so on Monday)

No problem.

> but using `NETOPT_IPV6_IID` is exactly what we don't want to do. It was deprecated in #10595. There are various reasons for that, but mainly we want to avoid code duplication, as not only the IID from a local device is needed, but also from remote devices (that's why we introduced `gnrc_netif_ipv6_iid_from_addr()` in the first place.

Ok, I see. Sorry, I didn't read all the related PRs and follow-ups, its just to complex for someone who is not involved :worried: My proposal was just to use something like the IID, no matter whether using `NETOPT_IPV6_IID` or function `gnrc_netif_ipv6_iid_from_addr`. In fact, the latter one is used when the netdev doesn't provide the the IID. When I was surfing through the code yesterday,  I already realized this code duplication in ESP-NOW and removed it already. I will provide a pull request.

So would `gnrc_netif_ipv6_iid_from_addr()` a possible option for generation and checking of ARO?

gschorcht commented:
> Also, the IID is not the interfaces EUI-64 (it refers to the IEEE 802.15.4 long address), as the second U/L-Bit gets still flipped.

I'm confused a bit, the IID as the second part of an IPv6 address is an EUI-64. If it is derived from MAC48, Bit 1 in first byte is already flipped. I would expect that `gnrc_netif_ipv6_iid_from_addr` provides exactly the EUI-64 from its MAC-48.

>  Maybe we should deactivate 6Lo-ND for ESP-now, as it is clearly only specified for IEEE 802.15.4m see abstract in RFC 6775.

Ok, the question is whether all other mechanisms required are then still working. I'm not able to asses this.

miri64 commented:
> > Also, the IID is not the interfaces EUI-64 (it refers to the IEEE 802.15.4 long address), as the second U/L-Bit gets still flipped.
> 
> I'm confused a bit, the IID as the second part of an IPv6 address is an EUI-64. If it is derived from MAC48, Bit 1 in first byte is already flipped. I would expect that `gnrc_netif_ipv6_iid_from_addr` provides exactly the EUI-64 from its MAC-48.

Yes and no. The [IID is formed from the EUI-64](https://tools.ietf.org/html/rfc2464#section-4) by flipping the bit, so the IID is semantically distinct from the IID (though RIOT-internally we use the same type for practicality, since syntactically they are the same). The EUI-64 for a IEEE 802.15.4 is given, as it [only knows EUI-64](https://tools.ietf.org/html/rfc4944#section-6) (and 2-bit short addresses were a stuffing algorithm is also specified). For a MAC-48 (or EUI-48) it needs to be formed as you described by stuffing. The main reason why I stated RFC 6775 might not apply for ESP-now (and most of our IEEE 802.15.4 devices, since we do not have an address provided by the vendor there) is that RFC 6775 assumes the EUI-64 is globally unique.

miri64 commented:
> > Maybe we should deactivate 6Lo-ND for ESP-now, as it is clearly only specified for IEEE 802.15.4m see abstract in RFC 6775.
> 
> Ok, the question is whether all other mechanisms required are then still working. I'm not able to asses this. There are a lot of places in the source code where it is distinguished between 6Lo and non-6Lo nodes. ESP-NOW is unfortunately acting like a 6Lo node.

Maybe #10499 helps with that? It adds a distinction between 6Lo-nodes and 6Lo-ND-nodes. I remember some IRC discussion with @TimoRoth that this might be necessary for ESP-now, but not the outcome :-/

gschorcht commented:
> Yes and no. The IID is formed from the EUI-64 by flipping the bit, so the IID is semantically distinct from the IID (though RIOT-internally we use the same type for practicality, since syntactically they are the same). 

Ah, I see, when you are talking about EUI-64 in data types and functions you are only referencing IEEE EUI-64 as is but not the modified EUI-64 as used for the IID in IPv6 address. This was no clear to me. Now, I got it.

> RFC 6775 assumes the EUI-64 is globally unique.

The OUIs used for an IEEE EUI-64 should be either the same if only 3 bytes are used or must not collide with an EUI-64 generated from EUI-48 if up to 36 bits are used. Therefore, an IEEE EUI-64 generated from an MAC address should be globally unique at least for devices with an official OUI in the MAC address.

So it should be theoretically possible for ESP-NOW nodes to generate a EUI-64 identifier by using function `eui48_to_ipv6_iid` and flipping back bit 1 in first byte. In that case, it would at least behave like a 6Lo node in ND.

gschorcht commented:
> Maybe #10499 helps with that? It adds a distinction between 6Lo-nodes and 6Lo-ND-nodes.

I guess, that multihop prefix feature will not work anymore since the border router and all routers if marked as non 6LNs are not 6LRs anymore and do not provide ABROs. Am I right?

miri64 commented:
> @miri64 Please add ` || DOXYGEN` here
> 
> [RIOT/sys/include/net/gnrc/netif.h](https://github.com/RIOT-OS/RIOT/blob/3af28369a8d49d7e18a0766ceb93098dd42c712e/sys/include/net/gnrc/netif.h#L80)
> 
> Line 80 in [3af2836](/RIOT-OS/RIOT/commit/3af28369a8d49d7e18a0766ceb93098dd42c712e)
> 
>  #if (GNRC_NETIF_L2ADDR_MAXLEN > 0) 
> to get it documented.

See https://github.com/RIOT-OS/RIOT/pull/10719

miri64 commented:
Rebased to current master

miri64 commented:
> * I had to comment out the following assertion, since the RA that is being received from Ethernet from LAN, of course, does not contain ABRO 
>     
>       
>         [RIOT/sys/net/gnrc/network_layer/ipv6/nib/nib.c](https://github.com/RIOT-OS/RIOT/blob/a971f29149ebe6c576e420439db3b9b34dfebe4f/sys/net/gnrc/network_layer/ipv6/nib/nib.c#L1444-L1447)
>       
>       
>           Lines 1444 to 1447
>         in
>         [a971f29](/RIOT-OS/RIOT/commit/a971f29149ebe6c576e420439db3b9b34dfebe4f)
>       
>       
>       
>       
>   
>           
>             
>              #if GNRC_IPV6_NIB_CONF_MULTIHOP_P6C 
>           
>   
>           
>             
>                          assert(abr != NULL);    /* should have been set in _handle_abro() */ 
>           
>   
>           
>             
>                          _nib_abr_add_pfx(abr, pfx); 
>           
>   
>           
>             
>              #endif  /* GNRC_IPV6_NIB_CONF_MULTIHOP_P6C */ 
>           
>       
>     
>   
>   . Might it be an option to change it as following:
>   ```diff
>   -       assert(abr != NULL);    /* should have been set in _handle_abro() */ 
>   -       _nib_abr_add_pfx(abr, pfx); 
>   +       if (abr != NULL) {    /* should have been set in _handle_abro() */
>   +          _nib_abr_add_pfx(abr, pfx);
>   +      }
>   ```

I don't understand this. I thought you deactivated `GNRC_IPV6_NIB_CONF_MULTIHOP_P6C`. So this `#ifdef` shouldn't even be hit. I agree however, that for a border router the assert might cause problems. Do you want to provide a fix or should I?.

miri64 commented:
> 3\. I could solve the last problem by adding the option `NETOPT_ADDRESS_LONG` to `esp_now_netdev`.

That sounds like an invitation for work-around (this is what I view your solution as) build on work-around. Let's rather fix that ARO function somehow. It might be needed for BLE. [RFC 7668](https://tools.ietf.org/html/rfc7668#section-3.2.4) says the generated EUI-64 shall be used in the case of the ARO, so I will do that.

> So, would gnrc_netif_ipv6_iid_from_addr() a possible option for generation and checking of ARO? If I understand it correctly, it doesn't matter which 8 bytes are in ARO, the only prerquiste is that the ARO sent and the ARO received are the same. Yes, according to the specification it should be the EUI-64.

I guess that answers also that comment.

> Function `_is_iface_eui64` insists on a size of 8 bytes

Will also fix that

miri64 commented:
> > I don't understand this. I thought you deactivated GNRC_IPV6_NIB_CONF_MULTIHOP_P6C. So this #ifdef shouldn't even be hit. I agree however, that for a border router the assert might cause problems. Do you want to provide a fix or should I?
> 
> Yes, I deactivated `GNRC_IPV6_NIB_CONF_MULTIHOP_P6C`. When I changed the `assert` to an `if` statement, it worked also without deactivating it explicitly. Therefore, the question was whether the `assert` is an absolute must or whether a `if` statement would be enough.

Ok thanks for clarifying. Since the proposal came from you I prefer if you did the PR for that.

gschorcht commented:
> Ok thanks for clarifying. Since the proposal came from you I prefer if you did the PR for that.

I was just asking since I'm not really sure and you are the expert with the inside knowledge. [RF6775, section 4.3](https://tools.ietf.org/html/rfc6775#section-4.3) says

> The ABRO MUST be included in all RA messages in the case when RAs are used to propagate information between routers (as described in Section 8.2).

which sounds like an `assert` is appropriate. I'm not really able to asset what side effects it would have when multihop prefix is not deactivated explicitly.

What is you opinion how a boarder router connected to a standard LAN and getting standard RAs should behave?

miri64 commented:
> What is you opinion how a boarder router connected to a standard LAN and getting standard RAs should behave?

On the standard LAN it must behave like a normal node (so non-6Lo-node). [Multihop prefix dissemination is only for 6Lo-ND](https://tools.ietf.org/html/rfc6775#section-8.1). I believe there is a mechanism described in [RFC 8505](https://tools.ietf.org/html/rfc8505) how to exchange information over the backbone, but I neither read that RFC yet nor is it implemented in RIOT at the moment.

miri64 commented:
(I get the feeling this discussion starts to get way out of sync with the actual topic of this PR. Shall we move elsewhere? I don't believe issues are the right place either though :-/)

miri64 commented:
> > The ABRO MUST be included in all RA messages in the case when RAs are used to propagate information between routers (as described in Section 8.2).
> 
> which sounds like an `assert` is appropriate. I'm not really able to asset what side effects it would have when multihop prefix is not deactivated explicitly.

The particular line line of code we are talking about is not 6LN-specific. We should make sure however, that non-6LR don't disseminate a prefix it received without multihop prefix dissemination. So how about 

```diff
diff --git a/sys/net/gnrc/network_layer/ipv6/nib/nib.c b/sys/net/gnrc/network_layer/ipv6/nib/nib.c
index 17bb9b9..abdb3e4 100644
--- a/sys/net/gnrc/network_layer/ipv6/nib/nib.c
+++ b/sys/net/gnrc/network_layer/ipv6/nib/nib.c
@@ -1415,7 +1415,8 @@ static uint32_t _handle_pio(gnrc_netif_t *netif, const icmpv6_hdr_t *icmpv6,
     if (pio->flags & NDP_OPT_PI_FLAGS_A) {
         _auto_configure_addr(netif, &pio->prefix, pio->prefix_len);
     }
-    if ((pio->flags & NDP_OPT_PI_FLAGS_L) || gnrc_netif_is_6lr(netif)) {
+    if ((pio->flags & NDP_OPT_PI_FLAGS_L) ||
+        (gnrc_netif_is_6lr(netif)) && (abr != NULL)) {
         _nib_offl_entry_t *pfx;
 
         if (pio->valid_ltime.u32 == 0) {
@@ -1442,8 +1443,9 @@ static uint32_t _handle_pio(gnrc_netif_t *netif, const icmpv6_hdr_t *icmpv6,
         if ((pfx = _nib_pl_add(netif->pid, &pio->prefix, pio->prefix_len,
                                valid_ltime, pref_ltime))) {
 #if GNRC_IPV6_NIB_CONF_MULTIHOP_P6C
-            assert(abr != NULL);    /* should have been set in _handle_abro() */
-            _nib_abr_add_pfx(abr, pfx);
+            if (abr != NULL) {
+                _nib_abr_add_pfx(abr, pfx);
+            }
 #endif  /* GNRC_IPV6_NIB_CONF_MULTIHOP_P6C */
             if (pio->flags & NDP_OPT_PI_FLAGS_L) {
                 pfx->flags |= _PFX_ON_LINK;
```

gschorcht commented:
Seems to be ok to me.

miri64 commented:
> > 3. I could solve the last problem by adding the option `NETOPT_ADDRESS_LONG` to `esp_now_netdev`.
> 
> That sounds like an invitation for work-around (this is what I view your solution as) build on work-around. Let's rather fix that ARO function somehow. It might be needed for BLE. [RFC 7668](https://tools.ietf.org/html/rfc7668#section-3.2.4) says the generated EUI-64 shall be used in the case of the ARO, so I will do that.
> 
> > So, would gnrc_netif_ipv6_iid_from_addr() a possible option for generation and checking of ARO? If I understand it correctly, it doesn't matter which 8 bytes are in ARO, the only prerquiste is that the ARO sent and the ARO received are the same. Yes, according to the specification it should be the EUI-64.
> 
> I guess that answers also that comment.
> 
> > Function `_is_iface_eui64` insists on a size of 8 bytes
> 
> Will also fix that

See https://github.com/RIOT-OS/RIOT/issues/10723 for some preliminary discussion.

miri64 commented:
> Seems to be ok to me.

See https://github.com/RIOT-OS/RIOT/pull/10724

Other testers: The discussion above had little to do with the actual PR (which is just moving functionality around) but rather to try to get ESP-now running.

maribu commented:
Btw: Adding

```sh
if ! lsmod | grep tun > /dev/null ; then
    sudo modprobe tun
fi
```

To the script run on `make term` would make the developers life easier ;-)

maribu commented:
OK, it does not work on the MSB-A2

This is what I did in `example/gnrc_border_router`:

```
make BOARD=msba2 PORT=/dev/ttyUSB0 flash
sudo modprobe tun
make BOARD=msba2 PORT=/dev/ttyUSB0 term
```

and in `examples/gnrc_networking` (while the border router was still running in a separate shell):

```
make BOARD=msba2 PORT=/dev/ttyUSB1 flash
make BOARD=msba2 PORT=/dev/ttyUSB1 term
```

But the MSB-A2 all get the same HW and IPv6 address. I changed the address on the board running `examples/gnrc_netowrking` after boot using:

```
ifconfig 7 del fe80::ff:fe00:22
ifconfig 7 add fe80::ff:fe00:42
ifconfig 7 set addr 42
```

miri64 commented:
@maribu what does not work on MSB-A2? Your comment doesn't really provide any information on that :-/

miri64 commented:
> Btw: Adding
>
> ```sh
> if ! lsmod | grep tun > /dev/null ; then
>     sudo modprobe tun
> fi
> ```
>
> To the script run on `make term` would make the developers life easier ;-)

I believe you are the first one to report such a problem. I (with both Ubuntu and Arch) never had to load that module (nor add it to `/etc/modules` or something like that).

maribu commented:
If I understood correctly, the boarder router should forward packets to the Internet. I did `ping6 <IPv6_ADDRESS>` and I only got a reply when I used the link lokal address of the border router, but not when pinging a global address. The output of `nib neigh` was empty at any time.

gschorcht commented:
@maribu I guess your border router is connected to your LAN. Pinging global like `2a00:1450:4013:c07::5e` from `dig AAAA www.google.de` requires that the way back is defined and you use a routable prefix for your 6LoWPAN. This would be for example the case when you use a subnet of the prefix that was assigned to your LAN from your ISP. Another test could be to define the route back to your border router on a machine in your LAN.

miri64 commented:
@maribu I tried on current master and there were two things I noticed when executing `make term` with the `gnrc_border_router` application on `msba2`:

1. it executes
    ```sh
    sudo sh /home/mlenders/Repositories/RIOT-OS/RIOT/dist/tools/ethos/start_network.sh /dev/ttyUSB0 tap0 2001:db8::/64 -tg -p "/dev/ttyUSB0"
    ```
    The last three parameters are wrong and don't belong there.
2. When executing 
    ```sh
    sudo sh /home/mlenders/Repositories/RIOT-OS/RIOT/dist/tools/ethos/start_network.sh /dev/ttyUSB0 tap0 2001:db8::/64
    ```
    I still need to toggle the reset pins by hand (I did it by starting `pyterm -tg -p /dev/ttyUSB0` and closing it immediately after). After that I can use the terminal of the border router (test with `help`)

After that I flashed and configured a second node (I removed lladdr `fe80::ff:fe00:22` and set the hwaddr to 23 and added `fe80::ff:fe00:23` and `2001:db8::ff:fe00:23`). I then added a route to the other node on the BR (`nib route add 7 2001:db8::/64 fe80::ff:fe00:23`) (instead I could also just compile the border router with `USEMODULE=gnrc_rpl`).

Interestingly enough I am able to ping `2001:db8::ff:fe00:23` from the Linux host, but not the other way around (I suspect because the global address never becomes valid which hints to a not working neighbor discovery / address registration).

I confirmed that this is also works for this branch the same way. Since this the bug I reported above also is on current master, but the addresses and MTU are still configured correctly in this branch I'd say it works as expected (address registration doesn't seem to work but also not in master, so I investigate).

miri64 commented:
> I confirmed that this is also works for this branch the same way. Since this the bug I reported above also is on current master, but the addresses and MTU are still configured correctly in this branch I'd say it works as expected (address registration doesn't seem to work but also not in master, so I investigate).

Same problem as described in #10723.

cgundogan commented on sys/include/net/gnrc/netif/internal.h:
```diff
@@ -403,7 +404,32 @@ static inline bool gnrc_netif_is_6lbr(const gnrc_netif_t *netif)
+ * @brief   Get the default link-layer address option for the given
+ *          gnrc_netif_t::device_type of a network interface
+ */
+netopt_t gnrc_netif_get_l2addr_opt(gnrc_netif_t *netif);
```
Aren't `@return` and `@param` missing here, or did you let them out intentionally?

miri64 replied:
Good point! I also fixed some weird behavior (and the `const`ness of parameter).

cgundogan commented:
addendum: please squash!

miri64 commented:
Squashed.
